mapassign_faststr: panic: runtime error: invalid memory address or nil pointer dereference

Hey. GLHF ^_^

`go version go1.17 linux/amd64`

My case:
```go
package main

import (
	"github.com/goccy/go-json"
)

var J = []byte(`{
  "a": {},
  "b": {},
  "c": {},
  "d": {},
  "e": {},
  "f": {},
  "g": {},
  "h": {
    "m": "1"
  },
  "i": {}
}`)

type T0 map[string]T2

type T2 struct {
	F0 string `json:"m"`
	F1 T3
}

type T3 struct {
	F0 string
	F1 T4
}

type T4 struct {
	F0 string
	F1 string
	F2 string
	F3 string
	F4 string
	F5 string
	F6 int
}

func main() {
	json.Unmarshal(J, &T0{})
}

```

Out:
```bash
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x1 pc=0x411213]

goroutine 1 [running]:
github.com/goccy/go-json/internal/decoder.(*mapDecoder).mapassign(0xc0000b6120, 0xc0000e6200, 0x0, 0x1, 0xc00009e430)
	/home/go/pkg/mod/github.com/goccy/go-json@v0.7.7/internal/decoder/map.go:49 +0x3c
github.com/goccy/go-json/internal/decoder.(*mapDecoder).Decode(0xc0000b6120, 0xc0000e6200, 0xc0000a0150, 0x5703c0, 0xc0000b4028)
	/home/go/pkg/mod/github.com/goccy/go-json@v0.7.7/internal/decoder/map.go:161 +0x550
github.com/goccy/go-json.unmarshal({0x5703c0, 0xc00009af70, 0x405759}, {0x4b7f80, 0xc0000b4028}, {0x0, 0x0, 0x4ba380})
	/home/go/pkg/mod/github.com/goccy/go-json@v0.7.7/decode.go:47 +0x187
github.com/goccy/go-json.Unmarshal(...)
	/home/go/pkg/mod/github.com/goccy/go-json@v0.7.7/json.go:276
main.main()
	oops/main.go:44 +0x79
exit status 2
```

--- a/decode_test.go
+++ b/decode_test.go
@@ -3732,3 +3732,46 @@ func TestDecodeBinaryTypeWithEscapedChar(t *testing.T) {
 		}
 	})
 }
+
+func TestIssue282(t *testing.T) {
+	var J = []byte(`{
+  "a": {},
+  "b": {},
+  "c": {},
+  "d": {},
+  "e": {},
+  "f": {},
+  "g": {},
+  "h": {
+    "m": "1"
+  },
+  "i": {}
+}`)
+
+	type T4 struct {
+		F0 string
+		F1 string
+		F2 string
+		F3 string
+		F4 string
+		F5 string
+		F6 int
+	}
+	type T3 struct {
+		F0 string
+		F1 T4
+	}
+	type T2 struct {
+		F0 string `json:"m"`
+		F1 T3
+	}
+	type T0 map[string]T2
+
+	var v T0
+	if err := json.Unmarshal(J, &v); err != nil {
+		t.Fatal(err)
+	}
+	if v["h"].F0 != "1" {
+		t.Fatalf("failed to assign map value")
+	}
+}

--- a/decode_test.go
+++ b/decode_test.go
@@ -3767,6 +3767,7 @@ func TestIssue282(t *testing.T) {
 	}
 	type T0 map[string]T2
 
+	// T2 size is 136 bytes. This is indirect type.
 	var v T0
 	if err := json.Unmarshal(J, &v); err != nil {
 		t.Fatal(err)

--- a/internal/decoder/map.go
+++ b/internal/decoder/map.go
@@ -9,26 +9,29 @@ import (
 )
 
 type mapDecoder struct {
-	mapType       *runtime.Type
-	keyType       *runtime.Type
-	valueType     *runtime.Type
-	stringKeyType bool
-	keyDecoder    Decoder
-	valueDecoder  Decoder
-	structName    string
-	fieldName     string
+	mapType                 *runtime.Type
+	keyType                 *runtime.Type
+	valueType               *runtime.Type
+	canUseAssignFaststrType bool
+	keyDecoder              Decoder
+	valueDecoder            Decoder
+	structName              string
+	fieldName               string
 }
 
 func newMapDecoder(mapType *runtime.Type, keyType *runtime.Type, keyDec Decoder, valueType *runtime.Type, valueDec Decoder, structName, fieldName string) *mapDecoder {
+	isStructIndirect := valueType.Kind() == reflect.Struct && runtime.IfaceIndir(valueType)
+	stringKeyType := keyType.Kind() == reflect.String
+	canUseAssignFaststrType := stringKeyType && !isStructIndirect
 	return &mapDecoder{
-		mapType:       mapType,
-		keyDecoder:    keyDec,
-		keyType:       keyType,
-		stringKeyType: keyType.Kind() == reflect.String,
-		valueType:     valueType,
-		valueDecoder:  valueDec,
-		structName:    structName,
-		fieldName:     fieldName,
+		mapType:                 mapType,
+		keyDecoder:              keyDec,
+		keyType:                 keyType,
+		canUseAssignFaststrType: canUseAssignFaststrType,
+		valueType:               valueType,
+		valueDecoder:            valueDec,
+		structName:              structName,
+		fieldName:               fieldName,
 	}
 }
 
@@ -45,8 +48,8 @@ func mapassign_faststr(t *runtime.Type, m unsafe.Pointer, s string) unsafe.Point
 func mapassign(t *runtime.Type, m unsafe.Pointer, k, v unsafe.Pointer)
 
 func (d *mapDecoder) mapassign(t *runtime.Type, m, k, v unsafe.Pointer) {
-	if d.stringKeyType {
-		mapV := mapassign_faststr(d.mapType, m, *(*string)(k))
+	if d.canUseAssignFaststrType {
+		mapV := mapassign_faststr(t, m, *(*string)(k))
 		typedmemmove(d.valueType, mapV, v)
 	} else {
 		mapassign(t, m, k, v)

--- a/internal/decoder/map.go
+++ b/internal/decoder/map.go
@@ -20,19 +20,29 @@ type mapDecoder struct {
 }
 
 func newMapDecoder(mapType *runtime.Type, keyType *runtime.Type, keyDec Decoder, valueType *runtime.Type, valueDec Decoder, structName, fieldName string) *mapDecoder {
-	isStructIndirect := valueType.Kind() == reflect.Struct && runtime.IfaceIndir(valueType)
-	stringKeyType := keyType.Kind() == reflect.String
-	canUseAssignFaststrType := stringKeyType && !isStructIndirect
 	return &mapDecoder{
 		mapType:                 mapType,
 		keyDecoder:              keyDec,
 		keyType:                 keyType,
-		canUseAssignFaststrType: canUseAssignFaststrType,
+		canUseAssignFaststrType: canUseAssignFaststrType(keyType, valueType),
 		valueType:               valueType,
 		valueDecoder:            valueDec,
 		structName:              structName,
 		fieldName:               fieldName,
 	}
+}
+
+const (
+	mapMaxElemSize = 128
+)
+
+// See detail: https://github.com/goccy/go-json/pull/283
+func canUseAssignFaststrType(key *runtime.Type, value *runtime.Type) bool {
+	indirectElem := value.Size() > mapMaxElemSize
+	if indirectElem {
+		return false
+	}
+	return key.Kind() == reflect.String
 }
 
 //go:linkname makemap reflect.makemap

